Hubble: fix unknown identities for some CIDR

rolinh opened the pull request:
Please, see individual commits for details.

Fixes: #11612

rolinh committed:
daemon: move hubble related methods to hubble.go

Let's keep everything hubble related in the same file. This commit does
not bring any functional changes.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
daemon: optimize LookupSecIDByIP and make it public

LookupSecIDByIP returns the security ID for a given IP. This commit
makes the method public as it is planned to be used by hubble's parser
as well. The signature is also slightly updated to not include an error
as it is totally possible to perform a lookup without generating
potential errors.

LookupSecIDByIP is also optimized in the following ways:

 - Handle the case where an IP could be <nil> (it *will* be the case
   when used by hubble).
 - Instead of immediately performing a search by CIDR, try looking up
   for the IP directly as the underlying ipcache lookup is more
   efficient (it's a straightforward hashmap access).
 - Use existing GetCIDRPrefixLengths method to get the prefixes.
 - Avoid IP+prefix --> string --> CIDR --> string conversion by directly
   creating the CIDR based on the IP and creating the appropriate mask.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

gandro commented on daemon/cmd/hubble.go:
```diff
@@ -158,3 +165,100 @@ func (d *Daemon) launchHubble() {
+		bits = net.IPv6len * 8
+	}
+	for _, prefixLen := range prefixes {
+		cidr := net.IPNet{
```
nit: You can optimize this further by check for `prefixLen == bits` and skip the lookup in that case (since that is a full IP lookup that you already did).

tklauser reviewed (approved):
Nit: the 3rd commit is lacking a subject line. It starts directly with `This commit updates hubble parser's IPGetter interface in 2 ways:`. Probably better to have a short summary of the commit as its subject line.

Otherwise LGTM regarding the agent changes. Didn't look in detail at the changes affecting `pkg/hubble`.

gandro commented on daemon/cmd/hubble.go:
```diff
@@ -158,3 +165,104 @@ func (d *Daemon) launchHubble() {
+			continue
+		}
+		cidr := net.IPNet{
+			IP:   ip,
```
This unfortunately does not work. The old code pushed the IP through `ParseCIDR` here to translate the IP (e.g. `192.0.2.1`) into the network prefix (e.g. `192.0.2.0`). 

The resulting `cidr.String()` in your version therefore will be `192.0.2.1/24` instead of `192.0.2.0/24`, which will case the lookup to fail.

rolinh replied:
Nice catch! I'll fix that.

rolinh committed:
daemon: fully implement hubble parser's IPGetter interface

This commit updates hubble parser's IPGetter interface in 2 ways:

 - Change the ip paramter type from string back to net.IP (as it was
   until recently).
 - Change the method name of LookupByIP to LookupSecIDByIP. This name is
   a bit more explicit and, more importantly, it matches the
   corresponding method name in cilium's daemon.

To implement these changes, cilium's daemon now includes a new method,
GetK8sMetadata, that wraps ipcache.GetK8sMetadata and takes a net.IP as
argument instead of a string as in the ipcache package variant.

With these changes, cilium's daemon now implements all of hubble
parser's getters interfaces. To make this more explicit, the
LookupSecIDByIP method has been moved to the hubble.go file along with
the other implementations.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

michi-covalent reviewed (approved):
neat

glibsm commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -201,7 +201,7 @@ type LookupEndpointIDByIPFunc func(ip net.IP) (endpoint *endpoint.Endpoint, err
 // LookupSecIDByIPFunc Func wraps logic to lookup an IP's security ID from the
 // ipcache.
 // See DNSProxy.LookupSecIDByIP for usage.
-type LookupSecIDByIPFunc func(ip net.IP) (secID ipcache.Identity, exists bool, err error)
```
interesting that err was never used
